api/service/model_endpoint_service.go

```diff
@@ -270,64 +271,80 @@ func (s *modelEndpointsService) createVirtualService(model *models.Model, endpoi
-
-	vs.Spec.Http = []*istiov1beta1.HTTPRoute{
-		&istiov1beta1.HTTPRoute{
-			Match: []*istiov1beta1.HTTPMatchRequest{
-				&istiov1beta1.HTTPMatchRequest{
-					Uri: &istiov1beta1.StringMatch{
-						MatchType: &istiov1beta1.StringMatch_Prefix{
-							Prefix: defaultMatchURIPrefix,
-						},
-					},
-				},
-			},
-			Rewrite: &istiov1beta1.HTTPRewrite{
-				Uri: versionEndpointPath,
-			},
-
-			Route:  httpRouteDestinations,
-			Mirror: mirrorDestination,
-		},
-	}
+	vs.Spec.Http = createHttpRoutes(versionEndpointPath, httpRouteDestinations, protocolValue)
 
 	return vs, nil
 }
 
-func (s *modelEndpointsService) parseModelEndpointHost(model *models.Model, versionEndpoint *models.VersionEndpoint) (string, error) {
-	veURL, err := url.Parse(versionEndpoint.URL)
-	if err != nil {
-		return "", errors.Wrapf(err, "failed to parse version endpoint url")
-	}
-
-	host := strings.Split(veURL.Hostname(), fmt.Sprintf(".%s.", model.Project.Name))
-
+// createModelEndpointHost create model endpoint host based on model name, project name, and domain inferred from versionEndpoint
+func (s *modelEndpointsService) createModelEndpointHost(model *models.Model, versionEndpoint *models.VersionEndpoint) (string, error) {
+	host := strings.Split(versionEndpoint.Hostname(), fmt.Sprintf(".%s.", model.Project.Name))
 	if len(host) != 2 {
 		return "", fmt.Errorf("invalid version endpoint url: %s. failed to split domain: %+v", versionEndpoint.URL, host)
 	}
 
 	domain := host[1]
+	return fmt.Sprintf("%s.%s.%s", model.Name, model.Project.Name, domain), nil
+}
 
-	modelEndpointHost := model.Name + "." + model.Project.Name + "." + domain
+// assignVersionEndpoint fetches destination version endpoints from database and assign to model endpoint.
+// assignVersionEndpoint validates version endpoint status and returns error if find no running version endpoint.
+func (c *modelEndpointsService) assignVersionEndpoint(ctx context.Context, endpoint *models.ModelEndpoint) (*models.ModelEndpoint, error) {
+	var protocolValue protocol.Protocol
+	for k := range endpoint.Rule.Destination {
+		versionEndpointID := endpoint.Rule.Destination[k].VersionEndpointID
 
-	return modelEndpointHost, nil
-}
+		versionEndpoint, err := c.versionEndpointStorage.Get(versionEndpointID)
+		if err != nil {
+			return nil, fmt.Errorf("version Endpoint with given `version_endpoint_id: %s` not found", versionEndpointID)
+		}
 
-func (s *modelEndpointsService) parseVersionEndpointPath(versionEndpoint *models.VersionEndpoint) (string, error) {
-	veURL, err := url.Parse(versionEndpoint.URL)
-	if err != nil {
-		return "", errors.Wrapf(err, "failed to parse version endpoint url")
+		if !versionEndpoint.IsRunning() {
+			return nil, fmt.Errorf("version Endpoint %s is not running, but %s", versionEndpoint.ID, versionEndpoint.Status)
+		}
+
+		// ensure that all version endpoint destination have same protocol
+		if protocolValue != "" && protocolValue != versionEndpoint.Protocol {
```

Comment:
Currently it's not possible since we have only 1 version endpoint for backing 1 model endpoint. This code is just for future case where we support traffic splitting. In that case, all of the version endpoint must have same protocol.